### ISSUE TITLE
inclusive design checklist

### ISSUE THREAD

**Issue by mouse-reeve** [bookwyrm-social/bookwyrm#525]
# Inclusive Web Design Checklist

https://github.com/Heydon/inclusive-design-checklist

- [ ] Minify CSS and JS, and remove unused/redundant code [[#1](https://developers.google.com/speed/docs/insights/MinifyResources)] 
- [ ] Maintain terse HTML, without over-reliance on `<div>` scaffolding [[#1](http://designingforperformance.com/optimizing-markup-and-styles/#divitis)] 
- [x] Use screen reader and keyboard accessible HTML [[#1](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML)] 
- [ ] Compress raster images [[#1](https://www.html5rocks.com/en/tutorials/speed/img-compression/)] 
- [ ] Optimize SVG path data [[#1](https://web-design-weekly.com/2014/10/22/optimizing-svg-web/)] 
- [x] Make sure heading levels describe a logical section/subsection structure [[#1](https://webaim.org/techniques/semanticstructure/)] 
- [ ] Only include heading elements where they introduce sections of content
- [x] Remove potentially insensitive or uninclusive language (use "singular they") [[#1](http://alexjs.com/)] 
- [ ] Give video content captions and transcripts
- [ ] Provide transcripts for audio content
- [x] Make sure main body (paragraph) text is no smaller than the default (user agent) size [[#1](https://www.smashingmagazine.com/2011/10/16-pixels-body-copy-anything-less-costly-mistake/)] 
- [x] Support "pinch zoom" (remove `user-scalable=no` if present)
- [x] Use relative units (`em`, `rem`, and `ch`), especially for font metrics
- [x] Make sure styles and scripts are not render blocking [[#1](https://csabapalfi.github.io/eliminate-render-blocking/)] 
- [ ] Install a service worker and cache all applicable assets [[#1](https://css-tricks.com/serviceworker-for-offline/)] 
- [x] Use content-based, not device-specific, media queries [[#1](http://bradfrost.com/blog/post/7-habits-of-highly-effective-media-queries/#content)] 
- [x] Provide alternatives and/or descriptions for complex visualizations
- [x] Include only clear, meaningful animations
- [x] Honor requests to remove animation via the `prefers-reduced-motion` media query
- [ ] Make sure controls do not elicit unexpected or jarring behavior [[#1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/consistent-behavior-receive-focus.html)]  [[#2](https://www.w3.org/TR/UNDERSTANDING-WCAG20/consistent-behavior-unpredictable-change.html)] 
- [x] Do not include third parties that compromise user privacy [[#1](https://css-tricks.com/potential-dangers-of-third-party-javascript/)] 
- [x] Do not recreate supported and expected browser behaviors with bespoke scripts
- [x] Support Windows high contrast mode (use images, not background images) [[#1](http://adrianroselli.com/2012/08/css-background-images-high-contrast-mode.html)] 
- [x] Provide alternative text for salient images [[#1](https://www.w3.org/WAI/tutorials/images/decision-tree/)] 
- [x] Apply `alt="` or `aria-hidden="true"` to decorative images [[#1](https://www.w3.org/WAI/tutorials/images/decorative/)] 
- [ ] Make sure text and background colors contrast sufficiently [[#1](https://accessibility.blog.gov.uk/2016/06/17/colour-contrast-why-does-it-matter/)] 
- [x] Provide `<title>`s that name the site and the specific page [[#1](https://www.w3.org/TR/WCAG20-TECHS/G88.html)] 
- [ ] Provide large touch "targets" for interactive elements [[#1](http://www.bbc.co.uk/guidelines/futuremedia/accessibility/mobile/design/touch-target-size)] 
- [x] Use data tables (`<table>`) for data only, not visual layout purposes
- [ ] Make scrollable elements focusable for keyboard users
- [x] Do not rely on color for differentiation of visual elements
- [x] Use the same design patterns to solve the same problems
- [x] Ensure keyboard focus order is logical regarding visual layout
- [ ] Lazy load large image assets
- [ ] Honour DNT (Do Not Track) header [[#1](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/DNT)] 
- [ ] Translate / spell out acronyms the first time you use them
- [x] Do not hijack standard scrolling behavior
- [x] Move focus between dialogs and the controls that invoked them
- [ ] Give all form elements permanently visible labels
- [x] Give grouped form elements group labels
- [x] Place labels above form elements
- [ ] Provide status and error messages as WAI-ARIA live regions
- [ ] Provide clear, unambiguous focus styles
- [x] Employ well-balanced, highly legible fonts (not too complex or elaborate)
- [x] Do not use very thin font faces [[#1](http://www.telegraph.co.uk/science/2016/10/23/internet-is-becoming-unreadable-because-of-a-trend-towards-light/)] 
- [x] Ensure states (pressed, expanded, invalid, etc) are communicated to assistive software
- [ ] Match semantics to behavior for assistive technology users
- [ ] Provide a default language and use `lang="[ISO code]"` for subsections in different languages
- [ ] Make controls look like controls; give them strong perceived affordance
- [ ] Underline links — at least in body copy
- [ ] Make sure all content belongs to a landmark element (`<header>`, `<footer>`, `<nav>`, `<main>`, etc)
- [ ] Avoid pure white or pure black shades
- [x] Mark invalid fields clearly and provide associated error messages
- [ ] Ensure content is not obscured through zooming (no fixed widths)
- [ ] Provide a `manifest.json` file for identifiable homescreen entries
- [x] Indicate swipe gesture support clearly, and provide simple tap-based alternatives
- [ ] Make sure data tables wider than their container can be scrolled horizontally
- [x] Avoid time constraints where possible; provide a clear warning and option to extend where not possible 
- [x] Label and describe the same things with the same terminology
- [x] Ensure disabled controls are not focusable
- [x] Do not instate "infinite scroll" by default; provide buttons to load more items
- [x] Avoid justified body text [[#1](https://www.w3.org/TR/WCAG20-TECHS/F88.html)] 
- [x] Provide enough spacing between lines of text (`line-height`) [[#1](https://www.w3.org/TR/WCAG20-TECHS/C21.html)] 
- [ ] Ensure PDF content is accessible (include tags) [[#1](https://webaim.org/techniques/acrobat/)] 
- [x] Provide a skip link if necessary  [[#1](https://webaim.org/techniques/skipnav/)] 
- [x] Avoid all-caps text [[#1](https://github.com/humanmade/hm-pattern-library/issues/75)] 
- [ ] Ensure that content is written as clearly and simply as possible [[#1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/meaning-supplements.html)] 
- [ ] Provide descriptive captions for figures
- [ ] Warn users of links that have unusual behaviors, like linking off-site, or loading a new tab
- [x] Make content easier to find and improve search results with structured data [[#1](https://developers.google.com/search/docs/guides/prototype)] 
- [ ] Use textual labels to make voice activation cues obvious
- [ ] Do not mark up subheadings/straplines with separate heading elements
- [x] Ensure primary calls to action are easy to recognize and reach
- [x] Avoid images of text — text that cannot be translated, selected, or understood by assistive tech
- [ ] Provide a print stylesheet (single column, with interactive content hidden)
- [x] Use well-established, therefore recognizable, icons and symbols
- [x] Subset fonts to just the characters needed
- [ ] Instead of obstructing users with CAPTCHAs, use honeypots [[#1](https://en.wikipedia.org/wiki/Honeypot_(computing))] 
- [ ] Begin long, multi-section documents with a table of contents
- [ ] Don"t make users perform actions to reveal content unless completely necessary
- [x] If content is meant to be hidden, ensure it is properly hidden to all users
- [x] Make sure controls within hidden content are not focusable
- [ ] Use `srcset` to tailor images to devices and reduce bandwidth costs
- [x] Do not auto focus form fields, on page load
- [ ] Break up long and complex forms into discrete sections and/or screens
- [ ] Make forms as short as possible; offer shortcuts like autocompleting the address using the postcode
- [x] Ensure the same content is available across different devices and platforms
- [ ] Inform the user when there are important changes to the application state
- [x] Make sure the purpose of a link is clearly described: "read more" vs. "read more about accessibility"

**Comment by rkingett** [comment]
Loving this! It may seem like a lot right now, but, this is just my script kitty opinion, code will become much easier to manage if these things are thought about in the design process. I can also try to look for a free bot that checks for almost all of these issues. 👍🏻
